### PR TITLE
Replace pass

### PR DESCRIPTION
**Pull request by Kuree**
This is necessary for power domain replacing normal mux in the interconnect.

### CODE REVIEW

File: gemstone/common/transform.py [hunk 1]
```diff
@@ -50,3 +50,48 @@ def or_reduction(gen: Generator, sub_circuit_name: str, signal_name: str,
     gen.wire(read_data_reduce_or.ports.O, pass_through)
 
     return gen.ports[input_name]
+
+
+def replace(parent: Generator, old_gen: Generator, new_gen: Generator):
```
Comment (rsetaluri) [inline gemstone/common/transform.py:55]: Great!

File: gemstone/common/transform.py [hunk 2]
```diff
@@ -50,3 +50,48 @@ def or_reduction(gen: Generator, sub_circuit_name: str, signal_name: str,
...
+    replace the old_gen with the new generator. The interfaces of
+    @old_gen has to be the same as the @new_gen
+
+    :param parent: parent generator that holds the old_gen
+    :param old_gen: target generator to be replaced
+    :param new_gen: new generator to use
+    :return: None
+    """
+    # we first make sure that the interfaces are the same
+    assert len(old_gen.ports) == len(new_gen.ports)
```
Comment (rsetaluri) [inline gemstone/common/transform.py:66]: we can simplify this to
```python
assert old_gen.ports.keys() == new_gen.ports.keys()
types = zip(*[[p.type() for p in gen.items()] for gen in (new_gen, old_gen)])
assert all([t[0] == t[1] for t in types])```

File: tests/common/test_transform.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
+from gemstone.common.transform import *
+from gemstone.generator import Const
+import magma
+import fault
+import tempfile
+
+
+def test_remove_simple():
```
Comment (rsetaluri) [inline tests/common/test_transform.py:8]: Good test. However, can we have the new generator be different? And then check that the original one has one semantics, and the one after `replace` has different?

**Review by rsetaluri — CHANGES_REQUESTED** [review]

File: gemstone/common/transform.py (hunk 2 above)
Comment (Kuree) [reply gemstone/common/transform.py:66]: As a not-python-power-user, this is very difficult for me to comprehend what's going on with the new code. I would prefer to keep it simple, stupid. On a philosophical sense,
> Everyone knows that debugging is twice as hard as writing a program in the first place. So if you're as clever as you can be when you write it, how will you ever debug it?
-- Brian Kernighan

File: tests/common/test_transform.py (hunk 3 above)
Comment (Kuree) [reply tests/common/test_transform.py:8]: Added more complex tests.

File: gemstone/common/transform.py (hunk 2 above)
Comment (rsetaluri) [reply gemstone/common/transform.py:66]: :D fair point. I'm ok with the verbose version

**Review by rsetaluri — APPROVED** [review]

**Pull request merged**